store/store_download_test.go

```diff
@@ -1252,3 +1253,280 @@ func (s *storeDownloadSuite) TestDownloadInfiniteRedirect(c *C) {
 	err := s.store.Download(s.ctx, "foo", targetFn, &snap.DownloadInfo, nil, s.user, nil)
 	c.Assert(err, ErrorMatches, fmt.Sprintf("Get %q: stopped after 10 redirects", mockServer.URL))
 }
+
+func (s *storeDownloadSuite) TestDownloadIconOK(c *C) {
+	const expectedName = "foo"
+	const expectedURL = "URL"
+	expectedContent := []byte("I was downloaded")
+
+	restore := store.MockDownloadIcon(func(ctx context.Context, name, etag, url string, w store.ReadWriteSeekTruncater) (string, error) {
+		c.Check(name, Equals, expectedName)
+		c.Check(url, Equals, expectedURL)
+		w.Write(expectedContent)
+		return "", nil
+	})
+	defer restore()
+
+	path := filepath.Join(c.MkDir(), "downloaded-file")
+	err := store.DownloadIcon(s.ctx, expectedName, path, expectedURL)
+	c.Assert(err, IsNil)
+	defer os.Remove(path)
+
+	c.Assert(path, testutil.FileEquals, expectedContent)
+}
+
+func (s *storeDownloadSuite) TestDownloadIconOKWithNewEtag(c *C) {
+	const expectedName = "foo"
+	const expectedURL = "URL"
+	expectedContent := []byte("I was downloaded")
+	const newEtag = "some-unique-value"
+
+	restore := store.MockDownloadIcon(func(ctx context.Context, name, etag, url string, w store.ReadWriteSeekTruncater) (string, error) {
+		c.Check(name, Equals, expectedName)
+		c.Check(etag, Equals, "")
+		c.Check(url, Equals, expectedURL)
+		w.Write(expectedContent)
+		return newEtag, nil
+	})
+	defer restore()
+
+	path := filepath.Join(c.MkDir(), "downloaded-file")
+	err := store.DownloadIcon(s.ctx, expectedName, path, expectedURL)
+	c.Assert(err, IsNil)
+	defer os.Remove(path)
+
+	c.Check(path, testutil.FileEquals, expectedContent)
+	etagBuf := make([]byte, 256)
+	size, err := unix.Getxattr(path, "user.etag", etagBuf)
+	c.Assert(err, IsNil)
+	writtenEtag := string(etagBuf[:size])
+	c.Check(writtenEtag, Equals, newEtag)
+}
+
+func (s *storeDownloadSuite) TestDownloadIconOKWithExistingEtag(c *C) {
+	const expectedName = "foo"
+	const expectedURL = "URL"
+	existingContent := []byte("I was already here")
+	responseContent := []byte("I should not be written")
+	const existingEtag = "some-unique-value"
+	path := filepath.Join(c.MkDir(), "downloaded-file")
+
+	// Create existing file
+	c.Assert(os.WriteFile(path, existingContent, 0o644), IsNil)
+	// Set etag xattr
+	c.Assert(unix.Setxattr(path, "user.etag", []byte(existingEtag), 0), IsNil)
```

Comment:
unit tests could be running on a system which does not support xattrs (very unlikely but still a possibility), in which case we should probably skip the test